ports/rp2/mphalport.c

```diff
@@ -44,6 +44,30 @@ ringbuf_t stdin_ringbuf = { stdin_ringbuf_array, sizeof(stdin_ringbuf_array) };
 
 #endif
 
+#if MICROPY_HW_ENABLE_USBDEV
+
+#ifndef USB_BUFFER_LEN
+#define USB_BUFFER_LEN 512
+#endif
+
+
+STATIC uint8_t usb_ringbuf_array[USB_BUFFER_LEN];
+ringbuf_t usb_ringbuf = { usb_ringbuf_array, sizeof(usb_ringbuf_array) };
+
+void tud_cdc_rx_cb(uint8_t itf) {
+    // consume pending USB data immediately so that the USB endpoint 
+    // does not get stalled.
+    for (uint32_t bytes_avail = tud_cdc_n_available(itf); bytes_avail > 0; bytes_avail --) {
+        if (ringbuf_free(&usb_ringbuf)) {
+            ringbuf_put(&usb_ringbuf, tud_cdc_read_char());
+        } else {
+            break;
```

Comment:
If this returns with bytes still available in the CDC buffer, and there's no more USB activity, I think those bytes will never be processed (because there's no further callback here, and `mp_hal_stdin_rx_chr` will only check in the ringbuf).

Reply:
I tried to check against this condition and while testing this with a bigger text buffer copy&pasted into the repl, I noticed that the board freezes (may happen after several chunks of say 600 bytes strings pasted into REPL). With freezing I mean that it no longer reacts on the REPL.

I then reverted back before I made the callback changes (current upstream master), and there it seems to work fine. Pasting many blocks of data in the REPL worked until a memoryError was thrown..

So I'm not convinced anymore if handling this callback is actually a good idea.. not until the source of the freeze is found.

Reply:
The problem could be because there is now some processing for each char in this callback, and that means the USB stack misses some IRQs.

This could be related to the fact that on rp2 the `tud_task()` is called from `MICROPY_VM_HOOK_POLL`, whereas on other tinyusb ports like mimxrt and samd, it's called directly from the USB IRQ (calling from `MICROPY_VM_HOOK_POLL` introduces some latency).

Given that those other ports work well calling `tud_task()` from IRQ context, it would be good to see if rp2 can do the same.

Reply:
hmm I just realized that `tud_int_handler()` does not seem to get called. At least I did not see any references to it. Or am I missing something here? Perhaps that would also explain some issues. 

I can try to include the USB IRQ handling (which would then call `tud_int_handler()` and `tud_task()`, in line with the Mimxrt and Samd port).

Reply:
I think the IRQ handler gets registered with the pico-sdk, see `lib/tinyusb/src/portable/raspberrypi/rp2040/dcd_rp2040.c`:
```
irq_set_exclusive_handler(USBCTRL_IRQ, dcd_rp2040_irq);
```

Reply:
Further digging and thoughts... 

https://github.com/hathach/tinyusb/issues/52  suggests that `tud_task()` should not be called in IRQ context.

The comment [here](https://github.com/hathach/tinyusb/blob/master/hw/bsp/rp2040/family.c#L195 ) discourages to install a custom USB irq for the rp2040

CircuitPython does seems to deregister the original IRQ and install a separate handler though (https://github.com/adafruit/circuitpython/blob/main/ports/raspberrypi/supervisor/usb.c#L37). Likely due to reducing the latency introduced when a USB irq is fired until tud_task() is called: A usb background task is scheduled in the irq, seems to be a CircuitPython addition there.

I guess we could try to do similar things and improve latency. Perhaps re-installing the IRQ and adding a callback as done by CircuitPython could be  beneficial. But how could we signal to call `tud_task()` from the IRQ?  Maybe re-use something from the machinery of `micropython.schedule()`?

On a general note, it does bother me that this is going to change a lot of things. And from all I've read so far, the existing approach how MicroPython handles the USB stack should be fine. It may not be very good in terms of latency, but it should still satisfy the requirments given by the tinyUSB stack. It's still unclear to me why the endpoint behaves the way it does in #7996... Perhaps it would really be better to report this upstream and get an opinion from there. Maybe this is a real bug and should be solved in the stack, rather than trying to circumfence it in the upper layers.

Reply:
> If this returns with bytes still available in the CDC buffer, and there's no more USB activity, I think those bytes will never be processed (because there's no further callback here, and mp_hal_stdin_rx_chr will only check in the ringbuf).

this is addressed now. There is a separate check in `mp_hal_stdio_poll`

Reply:
It looks like the `pending_cdc_bytes=0` at the end of this function will overwrite the `pending_cdc_bytes = bytes_avail` line.  Also it seems a waste to use 5 bytes for what is effectively a flag.  And also this won't work with multiple CDC interfaces (the `cdc_itf_nr` will be clobbered by the next interface if all are full).

I suggest the following: have a `uint8_t cdc_itf_pending` set of bits which are 1 if the corresponding interface needs attention.  Then in this function it'll be set like `cdc_itf_pending |= 1 << itf`, cleared like `cdc_itf_pending &= ~(1 << itf)` at the top of the function, and then in poll will be something like:
```c
for (uint8_t itf = 0; itf < MAX; ++itf) {
    if (cdc_itf_pending & (1 << itf)) {
        tud_cdc_rx_cb(itf);
    }
}
```

Reply:
> for (uint8_t itf = 0; itf < MAX; ++itf)
> 

Hmm what would `MAX` be? In the stm32 port, there is `MICROPY_HW_USB_CDC_NUM`, should that be reused here?